[MINOR][SQL][DOCS] Add notes of the deterministic assumption on UDF functions

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

Spark assumes that UDF functions are deterministic. This PR adds explicit notes about that.
## How was this patch tested?

It's only about docs.

dongjoon-hyun committed:
[SPARK-15282][SQL] UDF funtion is not always deterministic and need to be evaluated once.

SparkQA commented:
**[Test build #58530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58530/consoleFull)** for PR 13087 at commit [`43794e1`](https://github.com/apache/spark/commit/43794e17357bf0b9ea2a6a2a1864eb73aac2d206).

SparkQA commented:
**[Test build #58531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58531/consoleFull)** for PR 13087 at commit [`610a5b9`](https://github.com/apache/spark/commit/610a5b914844556377dc489ea6d853aeea5ec84b).

SparkQA commented:
**[Test build #58530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58530/consoleFull)** for PR 13087 at commit [`43794e1`](https://github.com/apache/spark/commit/43794e17357bf0b9ea2a6a2a1864eb73aac2d206).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58530/
Test FAILed.

SparkQA commented:
**[Test build #58531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58531/consoleFull)** for PR 13087 at commit [`610a5b9`](https://github.com/apache/spark/commit/610a5b914844556377dc489ea6d853aeea5ec84b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58531/
Test FAILed.

dongjoon-hyun commented:
There are several cases which assumes UDF is deterministic. It would be a big change to user. I'll revert the change on ScalaUDF, and update this PR to change optimizer not to duplicate the UDF expression.

SparkQA commented:
**[Test build #58535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58535/consoleFull)** for PR 13087 at commit [`ab5bc4b`](https://github.com/apache/spark/commit/ab5bc4b77ef2fc40684f680cde296f7d91e29aa2).

dongjoon-hyun commented:
The reported error scenario is the following.
```scala
scala> val df = sc.parallelize(Seq(("a", "b"), ("a1", "b1"))).toDF("old","old1")
scala> val udfFunc = udf((s: String) => {println(s"running udf($s)"); s })
scala> val newDF = df.withColumn("new", udfFunc(df("old")))
scala> val filteredOnNewColumnDF = newDF.filter("new <> 'a1'")
scala> filteredOnNewColumnDF.show
running udf(a)
running udf(a)
running udf(a1)
+---+----+---+
|old|old1|new|
+---+----+---+
|  a|   b|  a|
+---+----+---+
```
The result of this PR is like the following.
```scala
scala> filteredOnNewColumnDF.show
running udf(a1)
running udf(a)
+---+----+---+
|old|old1|new|
+---+----+---+
|  a|   b|  a|
+---+----+---+
```

SparkQA commented:
**[Test build #58535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58535/consoleFull)** for PR 13087 at commit [`ab5bc4b`](https://github.com/apache/spark/commit/ab5bc4b77ef2fc40684f680cde296f7d91e29aa2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58535/
Test PASSed.

dongjoon-hyun commented:
Hi, @liancheng and @cloud-fan .
This PR is similar with your commit, `[SPARK-13473][SQL] Don't push predicate through project with nondeterministic field(s)`.
This PR prevent pushing predicate through project with UDF function expression.
Could you review this PR?

SparkQA commented:
**[Test build #58647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58647/consoleFull)** for PR 13087 at commit [`afe5216`](https://github.com/apache/spark/commit/afe5216b9a713b45d4332929516715a7ccad6a5d).

SparkQA commented:
**[Test build #58647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58647/consoleFull)** for PR 13087 at commit [`afe5216`](https://github.com/apache/spark/commit/afe5216b9a713b45d4332929516715a7ccad6a5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58647/
Test PASSed.

SparkQA commented:
**[Test build #58874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58874/consoleFull)** for PR 13087 at commit [`294042d`](https://github.com/apache/spark/commit/294042df8b98fab2175da418c147986e8c45f654).

dongjoon-hyun commented:
Hi, @rxin .
Could you review this PR?

SparkQA commented:
**[Test build #58874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58874/consoleFull)** for PR 13087 at commit [`294042d`](https://github.com/apache/spark/commit/294042df8b98fab2175da418c147986e8c45f654).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58874/
Test PASSed.

linbojin commented:
Hi, @marmbrus
Could you review this PR?

rxin commented:
cc @cloud-fan

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1025,7 +1025,8 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
     case filter @ Filter(condition, project @ Project(fields, grandChild))
-      if fields.forall(_.deterministic) =>
+        if fields.forall(_.deterministic) &&
+          fields.forall(_.find(_.isInstanceOf[ScalaUDF]).isEmpty) =>
```
I'm not sure if I understand this correctly, do you mean `ScalaUDF` can be nondeterministic and we should always treat it as nondeterministic expression? If so, I think a better idea is just override `deterministic` in `ScalaUDF` and always return false.

dongjoon-hyun replied:
Thank you, @cloud-fan , again!
Yep. Right. Exactly, I really wanted to do that. So, I made my first initial commit for this PR.

But, you can see the result in the above.
* the first initial commit: https://github.com/apache/spark/pull/13087/commits/85fa0406280802259cd141d8ad9fdeb9b9173405

* Jenkins result: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58531/consoleFull

* My decision:

 > There are several cases which assumes UDF is deterministic. It would be a big change to user. I'll revert the change on ScalaUDF, and update this PR to change optimizer not to duplicate the UDF expression.

I still think that is a correct solution. I mean I totally agree with you. But, as you see, it needs to change other testsuites, so I thought I need commiters' decision to do that.

cloud-fan replied:
Can you look into it to see where we made this assumption? For now I prefer to override `deterministic` in `ScalaUDF`, if it's not a lot of effort to fix it.

dongjoon-hyun replied:
Great! No problem. I will try to fix other testsuite correctly.

SparkQA commented:
**[Test build #58969 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58969/consoleFull)** for PR 13087 at commit [`94ecdd4`](https://github.com/apache/spark/commit/94ecdd41ef0cfdf26ba4415611aa92e142dc9f51).

dongjoon-hyun commented:
According to @cloud-fan 's advice, the goal of this PR is now making `ScalaUDF` as a non-deterministic expression. Although this is a correct fix, one noticeable drawback is that we cannot use ScalaUDF in GROUP BY. I updated the description of this PR, too.

SparkQA commented:
**[Test build #58969 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58969/consoleFull)** for PR 13087 at commit [`94ecdd4`](https://github.com/apache/spark/commit/94ecdd41ef0cfdf26ba4415611aa92e142dc9f51).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58969/
Test FAILed.

dongjoon-hyun commented:
Hmm. @cloud-fan .
There is bad news. **ALS.scala** uses UDF on aggregation. So, there are 7 failures on **ALSSuite.scala**.
```scala
override def transform(dataset: Dataset[_]): DataFrame = {
    transformSchema(dataset.schema)
    // Register a UDF for DataFrame, and then
    // create a new column named map(predictionCol) by running the predict UDF.
    val predict = udf { (userFeatures: Seq[Float], itemFeatures: Seq[Float]) =>
      if (userFeatures != null && itemFeatures != null) {
        blas.sdot(rank, userFeatures.toArray, 1, itemFeatures.toArray, 1)
      } else {
        Float.NaN
      }
    }
    dataset
      .join(userFactors,
        checkedCast(dataset($(userCol)).cast(DoubleType)) === userFactors("id"), "left")
      .join(itemFactors,
        checkedCast(dataset($(itemCol)).cast(DoubleType)) === itemFactors("id"), "left")
      .select(dataset("*"),
        predict(userFactors("features"), itemFactors("features")).as($(predictionCol)))
}
```

According to the Jenkins test failure log, this is the last hurdle. However, it proves the usage of UDF on aggregation prevails. Spark users might depends on this risky feature much more.

dongjoon-hyun commented:
I updated `ALS` and `ALSSuite` just in order to pass the Jenkins for further discussion.

SparkQA commented:
**[Test build #58976 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58976/consoleFull)** for PR 13087 at commit [`1bf8b43`](https://github.com/apache/spark/commit/1bf8b43f2449bec1f9ce5d1a1952d4d5798fb2b3).

SparkQA commented:
**[Test build #58976 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58976/consoleFull)** for PR 13087 at commit [`1bf8b43`](https://github.com/apache/spark/commit/1bf8b43f2449bec1f9ce5d1a1952d4d5798fb2b3).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58976/
Test FAILed.

dongjoon-hyun commented:
The PySpark failure is fixed as a HOTFIX.

SparkQA commented:
**[Test build #59006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59006/consoleFull)** for PR 13087 at commit [`bb164ea`](https://github.com/apache/spark/commit/bb164ea76be81db78f2b4658a8ff0051aefee0ed).

marmbrus commented:
I think we can have a way to mark a UDF as non-deterministic, but that is too large of a change to make it the default.

Also, is this an actual performance problem, or does it just look like one (and common subexpression elimination is fixing it)?

markhamstra commented:
@marmbrus +1

dongjoon-hyun commented:
Thank you for review, @marmbrus and @markhamstra !

Actually, it's huge change. Although I'm not aware of the real background, the reported case can be handled by just preventing `PushDownPredicate` should not duplicate UDF function expressions. 

I think we can keep the common subexpression elimination without any change.

Anyway, I will revert the current investigation into my second commit back.

thunterdb commented:
@marmbrus +1 

I suggest to change the documentation of UDFs instead to underline the expectation that they be deterministic for the time being.

dongjoon-hyun committed:
Revert the change on ScalaUDF. Instead change PushDownPredicate optimizer.

dongjoon-hyun commented:
Thank you, @thunterdb .
By the way, do you mean we should resolve the JIRA issue as WONTFIX?

SparkQA commented:
**[Test build #59009 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59009/consoleFull)** for PR 13087 at commit [`9b81251`](https://github.com/apache/spark/commit/9b81251c1db1f70da8c9fcf6d35653ae9253c2eb).

dongjoon-hyun commented:
@marmbrus @markhamstra @thunterdb .
Now, this code and description of this PR is rollbacked my second commit 7 days ago.

For `common subexpression elimination`, I think it's okay if the some optimizer reduces the number of UDF calls. It's expectable. How do you think about this?

SparkQA commented:
**[Test build #59006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59006/consoleFull)** for PR 13087 at commit [`bb164ea`](https://github.com/apache/spark/commit/bb164ea76be81db78f2b4658a8ff0051aefee0ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59006/
Test PASSed.

SparkQA commented:
**[Test build #59009 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59009/consoleFull)** for PR 13087 at commit [`9b81251`](https://github.com/apache/spark/commit/9b81251c1db1f70da8c9fcf6d35653ae9253c2eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59009/
Test PASSed.

marmbrus commented:
My main questions remains: does this actually make a difference in runtime?  or is execution smart enough already to do this optimization (even if to the user it looks like its getting called more than once).

Also, @dongjoon-hyun would you have time to audit places where UDFs are documented and add the expectation that they are deterministic?

dongjoon-hyun commented:
For the documentation, sure, I will. We had better declare UDF should be deterministic as @thunterdb and you said from now.

For the runtime, I don't have any benchmark for this situation. Maybe, @linbojin , the reporter of this issue, might has some. @linbojin , could you answer to the @marmbrus 's question?

Generally speaking, this PR will prevent early pruning on all UDF filters for the semantics.
- Case 1: If the selectivity of UDF filter is high, this will cause performance degradation.
- Case 2: If the selectivity of UDF filter is low and the UDF takes long time, e.g. DB lookup or BLAS operation like `ALS.scala`, this will improve performance. As you see, the UDF will be called at every occurrence + 1.

 ```scala
val filteredOnNewColumnDF = newDF.filter("new <> 'a1' and new <> 'a2' and new <> 'a3'")
```

Although it depends, in most case, UDFs are assumed to have simple logic, and current `PushDownPredicate` is based on that assumption.

If Spark wants to just keep the current way to avoid any performance change, that's also okay for me. :) 
We can just declare explicitely that *ScalaUDF should be deterministic and might be called more than users declared*.

cloud-fan commented:
I like the idea to document `ScalaUDF` and say that the scala function must be deterministic. If users don't follow, they are on their own responsibility.

dongjoon-hyun commented:
Thank you. @cloud-fan .
By the way, to be clear with this, should we revert the change on `PushDownPredicate`, too?
I think it's another separate issue.

If the decision on this is finalized too, I can update this PR again.

Thank you for fast decision, @marmbrus , @markhamstra , @thunterdb , @cloud-fan .

linbojin commented:
@dongjoon-hyun @cloud-fan @marmbrus Thanks for your discussions about my reported issue: [SPARK-15282](https://issues.apache.org/jira/browse/SPARK-15282)

Maybe I should describe more about our use cases. In our project, firstly we generated a dataframe with one column called "fileName" one column called "url", and then we use a udf function (used inside withColumn()) to download the files from the corresponding urls and filter out '{}' data before writing to hdfs:

```scala
// df: DataFrame["fileName", "url"] 
val getDataUDF = udf((url: String) => {
	try { 
	   download data
	} catch { case e: Exception =>
      "{}"
    }
  })
val df2 = df.withColumn("data", getDataUDF(df("url")))
            .filter("data <> '{}'")
df2.write.save("hdfs path")
```

Based on our logs, each file will be downloaded twice. As for the running time, the writing job with filter will be twice as without filter: 
![screen shot 2016-05-22 at 22 19 24](https://cloud.githubusercontent.com/assets/5894707/15454461/c13c9918-206b-11e6-8901-1f473fbae3ca.png)
![screen shot 2016-05-22 at 22 18 02](https://cloud.githubusercontent.com/assets/5894707/15454462/c8d4c1a0-206b-11e6-88cc-8ad91d121b6e.png)
Left is with `.filter("data <> '{}'")` and right is without `.filter("data <> '{}'")`. It can be imaged, when there are many urls or the files are very large, the reported issue will affect the performance a lot.

Another problem is about data correctness. Because it's downloaded twice for each file, we came across some cases that the first downloading (getDataUDF) can get data (not '{}'), and the second downloading return '{}' because of certain connection exception. But i found the filter only worked on the first returned value so that spark will not remove this row but the value inside "data" column was '{}' which is the second returned value. Even after filter, we get the result dataframe df2 like the follows (files with '{}' data which should be removed):
```
fileName     url          data
file1        url1         sth     
files        url2         `{}`
```

**So on the high level, we get '{}' data after filter out '{}' which is strange. The reason I think is that UDF function is executed twice when filter on new column created by withColumn, and two returned values are different: first one makes filter condition true and second one makes filter condition false. The dataframe will keep the second value which in fact should not appear after filter operation.** 

Finally, i removed the filter operation (filter out '{}'  in downstream) because i think it may be not correct to  filter on new column created by withColumn. For me, i agree with @cloud-fan and @thunterdb, we can just document this behavior of udfs and uses should avoid to use udfs in such way.

dongjoon-hyun commented:
Thank you for sharing this, @linbojin .

dongjoon-hyun commented:
All issues are resolved now. I'll update this PR now.

dongjoon-hyun committed:
Revert code changes.

dongjoon-hyun committed:
Add notes on udf functions.

SparkQA commented:
**[Test build #59113 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59113/consoleFull)** for PR 13087 at commit [`7c11e2e`](https://github.com/apache/spark/commit/7c11e2e638e18759701c07826f4986be701ec591).

dongjoon-hyun commented:
I added for Scaladoc/Javadoc/Pydoc, but I cannot find proper places exposed in SparkR doc.

Hi, @shivaram , @davies , @felixcheung .
Should we need to add a note about `the assumption on UDF` in SparkR, too?

felixcheung commented:
@dongjoon-hyun I don't think we support UDF this way in SparkR?
Maybe something we could add to `dapply`? @sun-rui

dongjoon-hyun commented:
Thank you, @felixcheung ! Then, this PR is enough for the current master branch. :)
If SparkR have something related to this in the future, we can add a note on that at that time.

SparkQA commented:
**[Test build #59113 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59113/consoleFull)** for PR 13087 at commit [`7c11e2e`](https://github.com/apache/spark/commit/7c11e2e638e18759701c07826f4986be701ec591).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59113/
Test PASSed.

marmbrus commented on python/pyspark/sql/functions.py:
```diff
@@ -1756,6 +1756,7 @@ def __call__(self, *cols):
 @since(1.3)
 def udf(f, returnType=StringType()):
     """Creates a :class:`Column` expression representing a user defined function (UDF).
+    Note that the user-defined functions should be deterministic.
```
Thanks for doing this!  I would say `must be` deterministic.  We might even want to say that "due to optimization duplicate invocations may be eliminated or the function may even be invoked more times than it is present in the query".

dongjoon-hyun replied:
Thank you. I see. I'll fix like that.

dongjoon-hyun committed:
Address comments.

SparkQA commented:
**[Test build #59147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59147/consoleFull)** for PR 13087 at commit [`c1f92c7`](https://github.com/apache/spark/commit/c1f92c700b8de8bb9a4fc8116872544ed296d0d6).

dongjoon-hyun commented:
Hi, @marmbrus . 
I replaced 'should' with 'must', and added the detail description for `functions.py`, `SQLContext.scala`, `SparkSession.scala` and `UserDefinedFunction.scala`, too.

marmbrus commented:
LGTM pending tests.

@linbojin, we should also handle your use case though maybe that should be its own JIRA.  Perhaps you could open one with the information you posed here?

SparkQA commented:
**[Test build #59147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59147/consoleFull)** for PR 13087 at commit [`c1f92c7`](https://github.com/apache/spark/commit/c1f92c700b8de8bb9a4fc8116872544ed296d0d6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59147/
Test PASSed.

dongjoon-hyun commented:
Hi, @marmbrus .
Instead of creating new JIRA, I think we had better change the title of this PR into `[MINOR][SQL][DOC] ...`.
Initially, I tried to handle @linbojin 's SPARK-15282, but now this PR does not.
How do you think about that? Of course, @linbojin  can add the detail description to that JIRA issue.

marmbrus commented:
sure thats fine.

marmbrus commented:
merging to master and 2.0

dongjoon-hyun commented:
Oops. I didn't change the title yet.

dongjoon-hyun commented:
I'm not sure what happen. I'll remove this PR information from @linbojin 's JIRA issue anyway.

marmbrus commented:
Good thing I got distracted :)  I would have just changed the title while merging though.

dongjoon-hyun commented:
Thank you so much!

dongjoon-hyun commented:
Thank you all for reviewing and helping this PR!

linbojin commented:
@marmbrus @dongjoon-hyun I will add the detail description to the old SPARK-15282 JIRA issue.
